Tasks/Maven/mavencodeanalysis.ts

```diff
@@ -0,0 +1,231 @@
+        }
+    }
+    if (totalViolations == 1 && filesWithViolations == 1) {
+        // Looks like: 'PMD found 1 violation in 1 file.'
+        return tl.loc('codeAnalysisBuildSummaryLine_OneViolationOneFile', toolName);
+    }
+    if (totalViolations == 0) {
+        // Looks like: 'PMD found no violations.'
+        return tl.loc('codeAnalysisBuildSummaryLine_NoViolations', toolName);
+    }
+    // There should be no valid code reason to reach this point - '1 violation in 4 files' is not expected
+    throw new Error('Unexpected results from ' + toolName + ': '
+        + totalViolations + 'total violations in ' + filesWithViolations + ' files');
+}
+
+// Copy output files to a staging directory and upload them as build artifacts.
+// Each tool-module combination uploads its own build artifact.
+function uploadBuildArtifacts(toolName:string, moduleAnalysis:ma.ModuleAnalysis):void {
+    // The staging directory is used as a place to copy files before group uploading them, and was originally
+    // related to the following bug: https://github.com/Microsoft/vso-agent/issues/263
+    var localStagingDir:string = path.join(getStagingDirectory(toolName), moduleAnalysis.moduleName);
+    tl.mkdirP(localStagingDir);
+
+    var filesToUpload:string[] = [];
+    if (moduleAnalysis.analysisResults[toolName]) {
+        filesToUpload = moduleAnalysis.analysisResults[toolName].filesToUpload;
+    }
+
+    if (filesToUpload.length < 1) {
+        console.log('No artifacts to upload for ' + toolName + ' analysis of module ' + moduleAnalysis.moduleName);
+        return;
+    }
+
+    // Copy files to a staging directory so that they can all be uploaded at once
+    // This gives us a single artifact with all relevant files grouped together,
+    // giving a more organised experience in the artifact explorer.
+    filesToUpload.forEach((fileToUpload:string) => {
+        var stagingFilePath = path.join(localStagingDir, path.basename(fileToUpload));
+        tl.debug('Staging ' + fileToUpload + ' to ' + stagingFilePath);
+        // Execute the copy operation. -f overwrites if there is already a file at the destination.
+        tl.cp('-f', fileToUpload, stagingFilePath);
+    });
+
+    console.log('Uploading artifacts for ' + toolName + ' from ' + localStagingDir);
+
+    // Begin artifact upload - this is an asynchronous operation and will finish in the future
+    tl.command("artifact.upload", {
+        // Put the artifacts in subdirectories corresponding to their module
+        containerfolder: moduleAnalysis.moduleName,
+        // Prefix the build number onto the upload for convenience
+        // NB: Artifact names need to be unique on an upload-by-upload basis
+        artifactname: buildNumber + '_' + moduleAnalysis.moduleName + '_' + toolName
+    }, localStagingDir);
+}
+
+// Extract data from code analysis output files and upload results to build server
+export function uploadCodeAnalysisResults():void {
```

Comment:
I would have the "public" functions before the "private" functions to make the file easier to read, but I don't feel strongly about it.